Pre-lecture homework

Task 1
Using Dataset on Titanic from Tutorial Session:

In [15]:
import pandas as pd
url = "https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv"
titanic_data_with_missing = pd.read_csv(url)

# Check for missing data
missing_data = titanic_data_with_missing.isnull().sum()
print(missing_data)

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64


Task 2: Number of rows and columns in the dataset

In [16]:
titanic_df = pd.read_csv(url)
# Get the shape of the DataFrame (rows, columns)
rows, columns = titanic_df.shape

print(f'Number of rows: {rows}')
print(f'Number of columns: {columns}')

Number of rows: 891
Number of columns: 12


Task 2.2:

Observations refer to the individual data entries recorded for a single subject. One observation generally corresponds to one row in a dataset.
Variables refer to the characteristics that can be measured or counted for each observation, and are hence the columns in a dataset.

Task 3

In [18]:
# Summary statistics for numerical columns
print(df.describe())

       PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200  


Task 4

df.shape provides the total number of columns in the DataFrame, whereas df.describe() only summarises numeric columns by default and hence excludes columns of a non-numeric type. This leads to df.describe() showing less columns than df.shape for this particular dataset.

df.shape counts all entries for each column, including null entries which imply missing values. On the other hand, the "count" value reported by df.describe() reflects the number of non-null entries only. Any null entries, and hence missing values, are excluded. Thus, the values which df.describe reports in the "count" column are equal to the total count, as reported by df.shape, less the number of missing values for each particular column, if there are any.

Task 5

An attribute is a characteristic of the data being analysed, and gives more information about the DataFrame by simply retrieving values which are already stored. Meanwhile, a method is a function associated with an object that performs operations on the data stored in the object. 

_Post-Lecture Homework_

Task 6

The 'count' summary statistic gives the number of non-missing values for a variable. Missing values are excluded from the count.

'mean' gives arithmetic average of a variable. It is calculated by summing all the non-missing values and dividing by the count of non-missing values.

'std' refers to standard deviation, which measures the spread of values around the mean of a variable. The higher the standard deviation, the more dispersed the values are from the mean.

'min' gives the smallest value in the variable.

'25%' refers to the first quartile, or 25th percentile. It is the value below which 25% of the data points fall when the data points are arranged in ascending order. 

'50%' refers to the median, or 50th percentile. It is the middle value of a variable when the data points are arranged in ascending order. 

'75%' refers to the third quartile, or 75th percentile when the data points are arranged in ascending order.

'max' gives the largest value in the variable.

-------------------------------------------------------------------------------------------------------------------

ChatGPT transcript log (up to Question 6): https://chatgpt.com/share/4a1434f3-cac5-4c38-84ac-a871dfe9c372

ChatBot session summary (up to Question 6):
    
1. Loading and Inspecting the Dataset:

You downloaded a Titanic dataset from the URL: https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv.
We discussed how to find the number of rows and columns in the dataset using df.shape, which gives the total number of observations (rows) and variables (columns).

2. Definitions of Observations and Variables:

Observations: Rows in the dataset, each representing a unique passenger on the Titanic.
Variables: Columns in the dataset, representing different attributes of the passengers (e.g., age, fare, survival status).

3. Summarizing the Dataset:

We covered different methods for summarizing the dataset:
df.describe(): Provides summary statistics (count, mean, std, min, 25%, 50%, 75%, max) for numerical columns.
df.value_counts(): Shows the distribution of categorical variables.
df.info(): Displays the data types and the number of non-null values in each column.
df.isnull().sum(): Identifies missing values in each column.

4. Discrepancies Between df.shape and df.describe():

df.shape gives the total number of rows and columns, while df.describe() analyzes only the numerical columns by default.
The count values in df.describe() reflect non-missing entries, which is why they may differ from the total number of rows.

5. Difference Between Attributes and Methods:

Attributes (like df.shape) provide direct properties of the DataFrame without computation, hence no parentheses.
Methods (like df.describe()) perform operations or computations, and thus require parentheses when called.

6. Explanation of Summary Statistics:

Count: Number of non-null values in each column.
Mean: The average of the numerical data.
Standard Deviation (std): A measure of data spread around the mean.
Min, Max: Minimum and maximum values.
25%, 50% (Median), 75%: Percentiles (quartiles) dividing the data.

-------------------------------------------------------------------------------------------------------------------

Task 7

Part 1: 

Suppose we have a dataset about 100 patients in a hospital, which we would like to use to perform data analysis. We want to perform our analysis based only on observations with full data. Some variables (columns in the dataset) could include patient ID, name, age, gender, date of admission, room number and diagnosis. 

Now suppose 3 patients (observations) have missing entries for age, date of admission, room number and diagnosis. Using _df.dropna()_ will enable us to remove these 3 patients from the dataset, ensuring that our analysis is based only on patients with full data as we wanted. On the other hand, had we used _del df['diagnosis']_, for example, we would have removed information about the diagnosis of all patients, including the other 97 who did not have missing values for diagnosis. This would significantly inhibit our analysis.

Part 2:

Taking the same example of hospital patients, suppose we decided to include another variable, allergies. However, it is found that the vast majority of patients had no known allergies. Given that the 'allergies' column is mostly empty, it is deemed not important to our analysis (which requires observations with full data as assumed above). 

Using _del df['allergies']_ will allow us to delete the allergies column while keeping all the other variables such as patient ID, name, age, etc. intact. Meanwhile, using _df.dropna()_ in this instance would mean losing the data entries for all the patients who did not have any known allergies, and who make up the majority of the observations. We would lose a lot of data as a result.

Part 3:

Applying _del df['col']_ first allows us to remove variables which may not be relevant to our analysis, and which may contain missing values. By deleting these columns first, we are left with observations which have less missing values and therefore less rows will be removed from the dataset when applying _df.dropna()_. This helps us to conserve more data and reduce the risk of unnecessarily losing some data entries pertaining to more important variables.

Part 4: Consider the following dataset.

In [48]:
url = "https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv"
df = pd.read_csv(url)
df

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

In [49]:
df.isna().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

In [50]:
del df['Cabin']
del df['Age']
df.dropna(inplace=True)
df

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex  SibSp  Parch  \
0                              Braund, Mr. Owen Harris    male      1      0   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female      1      0   
2                               Heikkinen, Miss. Laina  female      0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female      1      0   
4                             Allen, Mr. William Henry    male      0      0   
..                                                 ...     ...    ...    ...   
886                              Montvila, Rev. Juozas    male      0      0   
887                       Graham, Miss. Margaret Edith  female      0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female      1      2   
889                              Behr, Mr. Karl Howell    male      0      0   
890                                Dooley, Mr. Patrick    male      0      0   

               Ticket     Fare Embarked  
0           A/5 21171   7.2500        S  
1            PC 17599  71.2833        C  
2    STON/O2. 3101282   7.9250        S  
3              113803  53.1000        S  
4              373450   8.0500        S  
..                ...      ...      ...  
886            211536  13.0000        S  
887            112053  30.0000        S  
888        W./C. 6607  23.4500        S  
889            111369  30.0000        C  
890            370376   7.7500        Q  

[889 rows x 10 columns]

In [51]:
df.isna().sum()

PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Embarked       0
dtype: int64

Explanation:

There were originally 687 missing values for 'Cabin' column, 177 for 'Age' and 2 for 'Embarked'. Given more than half of observations had missing values for 'Cabin', we delete the 'Cabin' column first. Next, 'Age' had 714 non-missing values which would be lost if the 'Age' column is removed. However, using _df.dropna()_ to remove the missing age values would mean losing data entries for 177 observations, which is more data than what is lost through _del df['Age']_. Hence, deleting the 'Age' column is more efficient. Lastly, only 2 values are missing for 'Embarked', and using _df.dropna()_ here would mean losing 2 observations which is better than losing the 889 non-missing data entries for 'Embarked'.

Task 8 

Part 1:

In [62]:
url = "https://raw.githubusercontent.com/KeithGalli/pandas/master/pokemon_data.csv"
df = pd.read_csv(url)
df

#                   Name   Type 1  Type 2  HP  Attack  Defense  \
0      1              Bulbasaur    Grass  Poison  45      49       49   
1      2                Ivysaur    Grass  Poison  60      62       63   
2      3               Venusaur    Grass  Poison  80      82       83   
3      3  VenusaurMega Venusaur    Grass  Poison  80     100      123   
4      4             Charmander     Fire     NaN  39      52       43   
..   ...                    ...      ...     ...  ..     ...      ...   
795  719                Diancie     Rock   Fairy  50     100      150   
796  719    DiancieMega Diancie     Rock   Fairy  50     160      110   
797  720    HoopaHoopa Confined  Psychic   Ghost  80     110       60   
798  720     HoopaHoopa Unbound  Psychic    Dark  80     160       60   
799  721              Volcanion     Fire   Water  80     110      120   

     Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0         65       65     45           1      False  
1         80       80     60           1      False  
2        100      100     80           1      False  
3        122      120     80           1      False  
4         60       50     65           1      False  
..       ...      ...    ...         ...        ...  
795      100      150     50           6       True  
796      160      110    110           6       True  
797      150      130     70           6       True  
798      170      130     80           6       True  
799      130       90     70           6       True  

[800 rows x 12 columns]

In [65]:
df.groupby("Type 2")["HP"].describe()

count       mean        std   min    25%   50%     75%    max
Type 2                                                                 
Bug         3.0  53.333333  18.929694  40.0  42.50  45.0   60.00   75.0
Dark       20.0  75.550000  18.173071  45.0  63.00  70.0   95.00  103.0
Dragon     18.0  82.166667  29.380366  40.0  66.25  79.0   90.00  150.0
Electric    6.0  88.166667  27.895639  50.0  71.25  87.5  106.75  125.0
Fairy      23.0  64.304348  28.893904  20.0  50.00  57.0   72.50  140.0
Fighting   26.0  79.461538  16.224009  48.0  65.75  80.0   91.00  110.0
Fire       12.0  71.250000  20.902914  45.0  53.75  67.5   88.75  100.0
Flying     97.0  71.391753  21.650903  30.0  55.00  70.0   85.00  150.0
Ghost      14.0  59.142857  23.144553   1.0  50.00  59.0   67.50  100.0
Grass      25.0  62.640000  16.790076  35.0  50.00  60.0   75.00  100.0
Ground     35.0  77.228571  23.716569  31.0  59.00  75.0  100.00  111.0
Ice        14.0  90.000000  30.707053  50.0  62.50  90.0  124.50  130.0
Normal      4.0  63.500000  17.233688  44.0  57.50  62.0   68.00   86.0
Poison     34.0  58.764706  17.381737  30.0  45.00  60.0   68.00  114.0
Psychic    33.0  72.212121  19.171524  30.0  60.00  70.0   86.00  105.0
Rock       14.0  68.071429  25.562786  20.0  51.00  65.0   78.50  115.0
Steel      22.0  64.636364  18.823859  25.0  58.50  70.0   70.75  110.0
Water      14.0  62.714286  22.956241  30.0  44.00  65.0   77.25  110.0

From the above example, Pokemons are grouped according to their 'Type 2'. .describe() then gives the summary statistics for the HP of the Pokemons in each 'Type 2' group: Bug, Dark, Dragon, ... 

Part 2:



The count value in df.describe() represents the number of non-missing values for each column across the entire dataset. On the other hand, the count value in _df.groupby("col1")["col2"].describe()_ has the primary function of showing the size and distribution of data across different groups in a dataset.

Part 3:

In [1]:
url = "https://raw.githubusercontent.com/KeithGalli/pandas/master/pokemon_data.csv"
df = pd.read_csv(url)
df

NameError: name 'pd' is not defined

In [2]:
import pandas as pd
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanics.csv"
df = pd.read_csv(url)
df

HTTPError: HTTP Error 404: Not Found

In [4]:
url = "https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv"
df = pd.read_csv(url)
Df.groupby("Survived")["Age"].describe()

NameError: name 'Df' is not defined

In [6]:
url = "https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv"
df = pd.read_csv(url
df

SyntaxError: '(' was never closed (80380618.py, line 2)

In [7]:
url = "https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv"
df = pd.read_csv(url)
df.group_by("col1")["col2"].describe()

AttributeError: 'DataFrame' object has no attribute 'group_by'

In [10]:
df.groupby("sex")["Age"].describe()

KeyError: 'sex'

In [12]:
df.groupby(Sex)["Age"].describe()

NameError: name 'Sex' is not defined

In [13]:
df.groupby("Sex")[Age].describe()

NameError: name 'Age' is not defined

Concluding comment:

In general, troubleshooting using ChatGPT has proven faster and more effective than through individual searches on Google. Upon presenting the error name to the ChatBot, a general explanation of the potential cause of the error is shown to the user, who can feed additional information about what the code trying to be executed was in order to receive a more precise solution to the problem. However, for longer pieces of code, the lack of contextual understanding displayed by ChatBots may lead to incorrect results. 

Troubleshooting through searches on Google tend to be slower due to the lack of accuracy regarding the search results. Online websites highlighting a particular error may contain general pieces of information useful in solving the coding error, but it is highly unlikely that a particular site both matches the specific code we are trying to execute and its associated error.

It is therefore plausible to say that working in a ChatBot session to fix the errors is faster and easier than using Google to search for and fix errors, though ChatBots should best be regarded as a complement rather than a subtitute to troubleshooting code. 

----------------------------------------------

ChatBot summary (Questions 7 onwards):

Summary of Exchanges:

Error Handling and Debugging:

NameError: name 'pd' is not defined: This error indicated that pandas had not been imported correctly. Solution: Ensure import pandas as pd is included in the script and that pandas is installed.

HTTPError: HTTP Error 404: Not Found: This error occurred when trying to access a dataset URL that was not found. Solution: Verify the URL or use the correct link to access the dataset.

NameError: name 'Df' is not defined: This error occurred due to incorrect capitalization of the DataFrame variable. Solution: Use consistent naming, such as df, and ensure correct capitalization.

SyntaxError: '(' was never closed: This error was due to a missing closing parenthesis. Solution: Ensure all parentheses are properly closed in the code.

AttributeError: 'DataFrame' object has no attribute 'group_by': This error was due to a typo in the method name. Solution: Use groupby instead of group_by to group DataFrame rows.

KeyError: 'sex': This error occurred because the column 'sex' did not exist in the DataFrame. Solution: Verify the column names with df.columns and use the correct column name.

NameError: name 'Age' is not defined: This error indicated that 'Age' was not a valid column name in the DataFrame. Solution: Verify the column names and use the correct ones from the dataset.

General Guidance:

For analyzing datasets, ensure that you correctly reference columns and methods according to their exact names and syntax.
Always check column names with df.columns to confirm their existence and correct usage.
Ensure that URLs are valid and accessible when loading datasets from the web.
Code Examples Provided:

Correct usage of pd.read_csv with proper URL and parentheses.
Example code for df.groupby("Type 1")["Attack"].describe() to demonstrate grouping and summarizing data.
Verification of column names with df.columns.

ChatBot transcript log (Question 7 onwards): https://chatgpt.com/share/66e38387-756c-8012-8d56-cf0502b98350